### PR TITLE
+ support for multi-language 3rd-party packages

### PR DESCRIPTION
**Pull request by ericprud**
tested with:
  node ./tools/build.js -t all plaintext
  node ./tools/build.js -t all plaintext shexc solidity sparql ttl

not tested for with test/detect or test/markup tests ('cause I don't
know how).

### CODE REVIEW

**Comment by yyyc514** [comment]
> not tested for with test/detect or test/markup tests ('cause I don't
know how).

If the build is working then just running the full test suite should ALSO test your submodules.  Is that not what you're seeing?

**Comment by yyyc514** [comment]
Oh wait this is a PR lol.... :-)  Let me take a closer look.

File: test/detect/index.js [hunk 1]
```diff
@@ -34,7 +34,7 @@ function testAutoDetection(language, {detectPath}) {
 
 describe('hljs.highlightAuto()', () => {
   before( async function() {
-    let thirdPartyLanguages = await getThirdPartyLanguages();
+    let thirdPartyPackages = await getThirdPartyPackages(); // !! not used
```
Comment (yyyc514) [inline test/detect/index.js]: It's used on line 49.  `listLanguages` in the build has the FULL language list.  We only need to figure out where the tests are at.

File: tools/lib/language.js [hunk 2]
```diff
@@ -9,7 +9,7 @@ const REQUIRES_REGEX = /\/\*.*?Requires: (.*?)\n/s
 const CATEGORY_REGEX = /\/\*.*?Category: (.*?)\n/s
 const LANGUAGE_REGEX = /\/\*.*?Language: (.*?)\n/s
 const {rollupCode} = require("./bundling.js")
-const { getThirdPartyLanguages } = require("./external_language")
+const { getThirdPartyPackages } = require("./external_language") // !! not used
```
Comment (yyyc514) [inline tools/lib/language.js]: Used on line 101.

**Comment by yyyc514** [comment]
Maybe look at your one failing test though:

```
  1) hljs.highlightAuto()
324       "before all" hook for "adding dynamic tests...":
325     Error: detectTestDir(1c) not found
```

File: test/detect/index.js (hunk 1 above)
Comment (ericprud) [reply test/detect/index.js]: oops, was being blind

File: tools/lib/language.js (hunk 2 above)
Comment (ericprud) [reply tools/lib/language.js]: here too

**Comment by ericprud** [comment]
i've been unable to run the tests; `test/api/ident.js` requires `../../build`. I've tried to `./tools/build.js -t cdn` and `all` and not seen an `index.js` show up. If I hack `ident.js` to require `../../build/highlight.js`, I get an undefined `hljs` in `build/highlight.js`. I gave up with that one.

**Comment by yyyc514** [comment]
You need a node build:

```
% git co squash_build_pipeline 
Already on 'squash_build_pipeline'
Your branch is up to date with 'origin/squash_build_pipeline'.
% npm install >
% node  --stack-size=65500  ./tools/build.js -t node  
Starting build.
Finished build.
Writing styles.
Writing package.json.
Writing languages.
................................................................................................................................................................................................
Writing highlight.js

% ./node_modules/.bin/mocha  -b false  --globals document test/

  hljs
    .IDENT_RE
...
```

**Comment by ericprud** [comment]
Weak. b0rk something there, didn't I?
notes for me to attack it in the morning:
```
git clone {git@github.com:highlightjs/,}highlight.js && cd $_
(mkdir extra && cd extra && git clone {git@github.com:highlightjs/,}highlightjs-shexc)
(git checkout squash_build_pipeline && npm install && node ./tools/build.js -t node)
./node_modules/.bin/mocha  -b false  --globals document test/
```
ends with yaml and shexc an passes 482.

```
git fetch origin +refs/pull/2336/merge && git merge FETCH_HEAD && node ./tools/build.js -t node
./node_modules/.bin/mocha  -b false  --globals document test/
```
passes 293, fails:
```
1) hljs.highlightAuto()
2) highlight() markup
```

This isn't the same error you got. Do you think I've screwed something up here?

File: test/detect/index.js (hunk 1 above)
Comment (yyyc514) [reply test/detect/index.js]: Can you remove the not used comments?

Comment (ericprud) [reply test/detect/index.js]: done.
Oh cool, the code snippit gets an \[Outdated] annotation above it!

File: tools/lib/language.js (hunk 2 above)
Comment (ericprud) [reply tools/lib/language.js]: done

**Comment by yyyc514** [comment]
Just FYI when you force push it's harder to review what you've changed...

**Comment by ericprud** [comment]
Re
> ```
> Error: detectTestDir(1c) not found
> ```
, I'd not noticed this in the top of `test/detect/index.js` which makes it OK to have no `detectPath`:
``` javascript
function testAutoDetection(language, {detectPath}) {
  const languagePath = detectPath || utility.buildPath('detect', language);
```
For all of the non-`extra` languages, `detectTestDir` fell through to `undefined` (now `null` to make it a bit more explicit).

> Just FYI when you force push it's harder to review what you've changed...

Oops, just pushed one I'd committed this AM. I've stopped generating them, though. (I can squash just before merge if you want.)

**Comment by ericprud** [comment]
should pass all tests with the `highlightjs-shexc/master` (482 tests) and `highlightjs-shexc/multi-lang` (484 tests).
last bug was `sparql.js` conflicting with `sas` 'cause it was too greedy with the keywords.

**Comment by yyyc514** [comment]
> now null to make it a bit more explicit).

Ok, that's fine. Though I don't much mind undefined being stand-in, could be my Ruby roots where we just have a single nil value.


> I can squash just before merge if you want.)

No need, GitHub does that automatically if I want.

> last bug was sparql.js conflicting with sas 'cause it was too greedy with the keywords.

LOL.  I can't wait until people start making PRs that change our code because their tests are failing in their 3rd party grammar, but perhaps that's just a bad dream. :)

**Pull request merged**